### PR TITLE
Fix retention and conversion defaults when switching

### PR DESCRIPTION
**Pull request by IstvanM**
This PR is a refactor that allowed me to set the default metric config values on the explore page when the metric type is changed.

It also contains some fixes on the model.py (about this I am not sure), I am sure there were issues. But they only could manifest in the notebook version.
Imho at one point we should get rid of the notebook helper methods in the model.py. It is very hard to maintain.

Now Retention or conversion metric without segments can be also configured.

image

image

### CODE REVIEW

File: mitzu/adapters/sqlalchemy_adapter.py [hunk 1]
```diff
@@ -245,7 +245,10 @@ def _generate_retention_series_cte(
             selects.append(index_sel)
             index += time_window.value
             curr_dt = curr_dt + time_window.to_relative_delta()
-
+        if len(selects) > 64:  # TODO: make this come from configs
+            raise ValueError(
+                "Too many retention periods to caluclate try reducing the scope"
```
Comment (IstvanM) [inline mitzu/adapters/sqlalchemy_adapter.py:250]: I am trying to prevent retention calculation for too many periods... Later this should come from a config.

File: mitzu/adapters/sqlalchemy_adapter.py [hunk 2]
```diff
@@ -675,7 +678,7 @@ def _get_segment_sub_query_cte(
         self,
         sub_query: SegmentSubQuery,
         group_field: Optional[M.EventFieldDef] = None,
-        resolution: Optional[M.TimeGroup] = None,
+        resolution: M.Resolution = M.Resolution.EVERY_EVENT,
```
Comment (IstvanM) [inline mitzu/adapters/sqlalchemy_adapter.py:681]: Every Event resolution was NONE... this was very meh

File: mitzu/model.py [hunk 3]
```diff
@@ -1113,19 +1118,45 @@ def get_id(self) -> str:
 DEF_TIME_GROUP = TimeGroup.DAY
 
 
+class Resolution(Enum):
```
Comment (IstvanM) [inline mitzu/model.py:1121]: We used timegroup as resolution before, however that didn't really cover all events. Also it allowed YEAR, MONTH, WEEK resolution types as well... which is not ideal

File: mitzu/model.py [hunk 4]
```diff
@@ -1113,19 +1118,45 @@ def get_id(self) -> str:
...
+
+
 @dataclass()
 class MetricConfig:
     start_dt: Optional[datetime] = None
     end_dt: Optional[datetime] = None
-    lookback_days: Optional[TimeWindow] = None
-    time_group: Optional[TimeGroup] = None
-    max_group_count: Optional[int] = None
+    lookback_days: TimeWindow = DEF_LOOK_BACK_DAYS
```
Comment (IstvanM) [inline mitzu/model.py:1151]: This change probably should be a different PR, I started to move the default values out of the Metric. This is not necessarily the best place either. However the Metric should be a simple DTO at one point.... now it had a lot of defaults creating logic (because the notebook version required it)

File: mitzu/webapp/cache.py [hunk 5]
```diff
@@ -39,12 +39,22 @@ def list_keys(
...
-        default_factory=lambda: diskcache.Cache(configs.DISK_CACHE_PATH)
-    )
+    _disk_cache: diskcache.Cache
+
+    def __init__(self, name: str) -> None:
+        super().__init__()
+        object.__setattr__(
+            self,
+            "_disk_cache",
+            diskcache.Cache(timeout=10, directory=f"{configs.DISK_CACHE_PATH}/{name}"),
```
Comment (IstvanM) [inline mitzu/webapp/cache.py:52]: I am separating the CACHE from the DASH - QUEUE... It increases resiliency slightly

File: mitzu/webapp/pages/explore/explore_page.py [hunk 6]
```diff
@@ -185,7 +182,21 @@ def create_explore_page(
...
+            ret_or_conv_window = metric._conv_window
+        elif isinstance(metric, M.RetentionMetric):
+            ret_or_conv_window = metric._retention_window
+    else:
+        if metric_type == MTH.MetricType.CONVERSION:
+            ret_or_conv_window = M.DEF_CONV_WINDOW
+        elif metric_type == MTH.MetricType.RETENTION:
+            ret_or_conv_window = M.DEF_CONV_WINDOW
+
+    metrics_config_card = MC.from_metric(metric_config, metric_type, ret_or_conv_window)
```
Comment (IstvanM) [inline mitzu/webapp/pages/explore/explore_page.py:199]: The main change here is the core of the REFACTOR. Previously I was creating a Metric from all the DASH inputs. This was a bit problematic, as if Conversion didn't have any Segments we needed to return NONE. But because of the NONE the explore page always fell back to the DEFAULT on the metric configuration side as well. (On the right side). Now we can change any input component on the metric page and it will stay there.

File: mitzu/webapp/pages/explore/metric_config_handler.py [hunk 7]
```diff
@@ -250,28 +274,64 @@ def from_all_inputs(
...
-    res_tw = M.TimeWindow(
-        value=all_inputs.get(TIME_WINDOW_INTERVAL, 1),
-        period=M.TimeGroup(all_inputs.get(TIME_WINDOW_INTERVAL_STEPS, M.TimeGroup.DAY)),
-    )
-
-    res_val = all_inputs.get(RESOLUTION_DD, EVERY_EVENT_RESOLUTION)
-    resolution: Optional[M.TimeGroup] = None
-    if res_val != EVERY_EVENT_RESOLUTION:
-        resolution = M.TimeGroup.parse(res_val)
+    if ctx.triggered_id == MTH.METRIC_TYPE_DROPDOWN:
```
Comment (IstvanM) [inline mitzu/webapp/pages/explore/metric_config_handler.py:277]: If we change the METRIC type dropdown on the top of the page. We should set different default values. Otherwise the app could freeze if we switch to retention from a very long time-horizon segmentation... Same goes for Conversion. So this change just prevents the app freezing.

File: mitzu/model.py [hunk 8]
```diff
@@ -1113,19 +1118,45 @@ def get_id(self) -> str:
 DEF_TIME_GROUP = TimeGroup.DAY
 
 
+class Resolution(Enum):
+    EVERY_EVENT = auto()
+    ONE_USER_EVENT_PER_MINUTE = auto()
+    ONE_USER_EVENT_PER_HOUR = auto()
+    ONE_USER_EVENT_PER_DAY = auto()
```
Comment (medgyes) [inline mitzu/model.py:1125]: I'm wondering is we could make it a bit more automatic like how elastic search works. One sets a time range and the elastic will figure out the resolution balancing between useful charts and estimated resources for calculations

File: mitzu/model.py [hunk 9]
```diff
@@ -1322,57 +1347,53 @@ def config(
         start_dt: Optional[str | datetime] = None,
         end_dt: Optional[str | datetime] = None,
         custom_title: Optional[str] = None,
-        retention_window: Optional[Union[TimeWindow, str]] = None,
-        time_group: Optional[TimeGroup] = None,
+        retention_window: Union[TimeWindow, str] = TimeWindow(
```
[the remark this answers is not in the document]
Comment (medgyes) [reply mitzu/model.py:1350]: Here the default value isn't a reference to an instance? As I remember this can be dangerous if the rentention_window's value changes then the default value changes too

File: mitzu/model.py [hunk 10]
```diff
@@ -1113,19 +1118,45 @@ def get_id(self) -> str:
...
+    lookback_days: TimeWindow = DEF_LOOK_BACK_DAYS
+    time_group: TimeGroup = DEF_TIME_GROUP
+    max_group_count: int = DEF_MAX_GROUP_COUNT
     group_by: Optional[EventFieldDef] = None
     custom_title: Optional[str] = None
     agg_type: Optional[AggType] = None
     agg_param: Optional[Any] = None
     chart_type: Optional[SimpleChartType] = None
-    resolution: Optional[TimeGroup] = None
+    resolution: Resolution = Resolution.EVERY_EVENT
```
Comment (medgyes) [inline mitzu/model.py:1159]: wouldn't it be super slow to deal with every event by default?

**Review by medgyes — COMMENTED** [review]
I haven't reviewed the entire PR, just parts of it

File: mitzu/model.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (IstvanM) [reply mitzu/model.py:1350]: Time window is a frozen dataclass. That is why it is safe. Also it doesn't have anything mutable inside it like a List.
This is partially why frozen classes are good. Btw your point is valid, in case the data-class is not frozen.

File: mitzu/model.py (hunk 8 above)
Comment (IstvanM) [reply mitzu/model.py:1125]: I am kind of doing that in this PR. But based so far only on the type of the metric. Retention has hour resolution, conversion has minute resolution. But indeed we can do that. 

Once we have an `explore_service` that can be done.

File: mitzu/model.py (hunk 10 above)
Comment (IstvanM) [reply mitzu/model.py:1159]: In case of segmentation and conversion most often it is very fast. (Even shapr3d size data). By default, I think we shouldn't make assumptions on the model level. At services level we can make those assumptions. I wanted to keep this none-optional.

File: mitzu/webapp/pages/explore/explore_page.py [hunk 11]
```diff
@@ -185,7 +182,21 @@ def create_explore_page(
 
 
 def create_graph_container(metric: Optional[M.Metric], project: M.Project):
-    metrics_config_card = MC.from_metric(metric)
+    metric_config = metric._config if metric is not None else M.MetricConfig()
+    metric_type = MTH.MetricType.from_metric(metric)
```
Comment (medgyes) [inline mitzu/webapp/pages/explore/explore_page.py:186]: note for myself: this metric type is not the metric type from the models.py and if the metric arg is None then it's a segmentation metric

File: mitzu/webapp/pages/explore/graph_handler.py [hunk 12]
```diff
@@ -175,7 +181,6 @@ def create_callbacks():
             metric_id=State(EXP.METRIC_ID_VALUE, "children"),
         ),
         interval=configs.GRAPH_POLL_INTERVAL_MS,
-        prevent_initial_call=True,
```
Comment (medgyes) [inline mitzu/webapp/pages/explore/graph_handler.py:178]: is it intentional? wouldn't it put an extra load on the db on every page load?

**Review by medgyes — DISMISSED** [review]
